### PR TITLE
Show live code coverage as the app runs

### PR DESCRIPTION
**Pull request by devoncarew**
- add a `FlutterAppManager` project singleton, which clients can use to get events as the active running flutter app changes (it uses the current focused app from the POV of the Run or Debug pane)
- add a `FlutterLiveCoverageManager`, which listens for app runs, and shows and tracks live code coverage for the current file and app
- have a compile time constant (`ENABLE_LIVE_COVERAGE = false`) to enable the feature
- update coverage when:
  - the file changes
  - the user reloads or restarts
  - a pause event is hit
  - a frame is painted
- show covered lines using a flutter blue color, and non-covered lines using a gray color; the covered markers blend together, and the non-covered markers stay separate. As you cover lines, the left ruler paints in with blue markers

Some outstanding work:
- handle the case where the files modification date is newer than what's running on the VM (we'd be applying the VM's code coverage info to a file that the user has since edited); we should do something here like just not apply updated coverage info until the next reload
- valid our current UX, and decide when to enable behind a user visible flag

screen shot 2018-04-16 at 1 36 41 pm

screen shot 2018-04-16 at 1 37 02 pm

### CODE REVIEW

**Comment by devoncarew** [comment]
And generally:
- `FlutterAppManager` provides events when the current flutter app changes
- `FlutterLiveCoverageManager` manages any active live coverage sessions for a project - it only allows one active session, for the current active flutter app
- `FlutterAppLiveCoverage` manages a coverage session for a single app
- `EditorCoverageDecorations` manages coverage decorations for a single file

File: src/io/flutter/FlutterInitializer.java [hunk 1]
```diff
@@ -168,6 +170,9 @@ public void runActivity(@NotNull Project project) {
 
     FlutterRunNotifications.init(project);
 
+    // Start the live coverage manager.
+    FlutterLiveCoverageManager.getInstance(project);
```
Comment (pq) [inline src/io/flutter/FlutterInitializer.java]: Nit: consider renaming this to `init(...)` ?

File: src/io/flutter/actions/ReloadFlutterApp.java [hunk 2]
```diff
@@ -48,7 +52,16 @@ public void actionPerformed(AnActionEvent e) {
...
       // Else perform a hot reload.
       FlutterInitializer.sendAnalyticsAction(this);
-      FlutterReloadManager.getInstance(project).saveAllAndReload(getApp());
+      final Editor editor = FileEditorManager.getInstance(project).getSelectedTextEditor();
+
+      VirtualFile file = editor == null ? null : FileDocumentManager.getInstance().getFile(editor.getDocument());
+      if (file != null && !file.isInLocalFileSystem()) {
+        file = null;
+      }
+      if (file != null && !file.getPath().endsWith(".dart")) {
```
Comment (pq) [inline src/io/flutter/actions/ReloadFlutterApp.java]: Maybe `file != null && FlutterUtils.isDartFile(file)`)?

File: src/io/flutter/coverage/EditorCoverageDecorations.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.awt.*;
+import java.util.ArrayList;
+import java.util.List;
+
+class EditorCoverageDecorations implements Disposable {
+  private static final VirtualFileManager virtualFileManager = VirtualFileManager.getInstance();
+
+  private static final int HIGHLIGHTERLAYER = HighlighterLayer.SELECTION - 1;
```
[the remark this answers is not in the document]
Comment (pq) [reply src/io/flutter/coverage/EditorCoverageDecorations.java]: 👍

File: src/io/flutter/coverage/EditorCoverageDecorations.java [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+      final ScriptRef scriptRef = scripts.get(reportRange.getScriptIndex());
+      final String uri = scriptRef.getUri();
+      if (uri.startsWith("file:")) {
+        final VirtualFile file = virtualFileManager.findFileByUrl(uri);
+        if (file != null && file.equals(reloadFile)) {
+          scriptManager.populateFor(scriptRef);
+
+          //final int startIndex = reportRange.getStartPos();
+          //final int endIndex = reportRange.getEndPos();
```
Comment (pq) [inline src/io/flutter/coverage/EditorCoverageDecorations.java]: Is this still on deck or can we remove?

File: src/io/flutter/coverage/FileCoverageInfo.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      uncoveredLines.add(line);
+    }
+  }
+
+  public void addUncovered(Pair<Integer, Integer> pos) {
+    if (pos == null) {
+      return;
+    }
+    final int line = pos.first;
+    if (!coveredLines.contains(line)) {
```
Comment (pq) [inline src/io/flutter/coverage/FileCoverageInfo.java]: Maybe replace with a call to `addUncovered(line)` or `addUncovered(pos.first)`?

File: src/io/flutter/coverage/FileCoverageInfo.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      return;
+    }
+
+    final int line = pos.first;
+    coveredLines.add(line);
+    uncoveredLines.remove(line);
+  }
+
+  public void addUncovered(int line) {
+    if (!coveredLines.contains(line)) {
```
Comment (pq) [inline src/io/flutter/coverage/FileCoverageInfo.java]: => `isCovered(line)`?

File: src/io/flutter/coverage/FlutterAppLiveCoverage.java [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+  }
+
+  /**
+   * Schedule a repaint of the coverage information.
+   * <p>
+   * When.now schedules a repaint immediately.
+   * <p>
+   * When.soon will schedule a repaint shortly; that can get delayed by another request, with a maximum delay.
+   */
+  private void requestRepaint(When when) {
```
Comment (pq) [inline src/io/flutter/coverage/FlutterAppLiveCoverage.java]: Is the `when` being used in this call?

File: src/io/flutter/coverage/FlutterAppLiveCoverage.java [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+    if (!isConnected() || this.currentFile == null || this.currentEditor == null) {
+      return;
+    }
+
+    requestInProgress = true;
+
+    final FileEditor editor = this.currentEditor;
+    final VirtualFile file = this.currentFile;
+
+    JobScheduler.getScheduler().schedule(() -> performRequest(editor, file), 0, TimeUnit.SECONDS);
```
Comment (pq) [inline src/io/flutter/coverage/FlutterAppLiveCoverage.java]: I was expecting the `when` to get used here...

File: src/io/flutter/coverage/FlutterAppLiveCoverage.java [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+    if (isDirty) {
+      requestRepaint(When.soon);
+    }
+  }
+
+  public void showFor(@Nullable VirtualFile file, @Nullable FileEditor fileEditor) {
+    this.currentFile = file;
+    this.currentEditor = fileEditor;
+
+    // harvest old editors
```
Comment (pq) [inline src/io/flutter/coverage/FlutterAppLiveCoverage.java]: => Harvest old editors.

?

File: src/io/flutter/coverage/ScriptManager.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+      final String uri = libraryRef.getUri();
+
+      if (uri.startsWith("file:")) {
+        final VirtualFile libraryFile = virtualFileManager.findFileByUrl(uri);
+
+        if (file.equals(libraryFile)) {
+          final Library library = getLibrary(libraryRef);
+          if (library != null) {
+            if (!library.getScripts().isEmpty()) {
+              return library.getScripts().get(0);
```
Comment (pq) [inline src/io/flutter/coverage/ScriptManager.java]: Should we do anything if there is more than one?

File: src/io/flutter/coverage/ScriptManager.java [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+  public void populateFor(ScriptRef scriptRef) {
+    if (!scriptMap.containsKey(scriptRef.getId())) {
+      scriptMap.put(scriptRef.getId(), getScriptSync(scriptRef));
+      linesAndColumnsMap.put(scriptRef.getId(), createTokenPosToLineAndColumnMap(scriptMap.get(scriptRef.getId())));
+    }
+  }
+
+  public Pair<Integer, Integer> getLineColumnPosForTokenPos(ScriptRef scriptRef, int tokenPos) {
+    final TIntObjectHashMap<Pair<Integer, Integer>> map = linesAndColumnsMap.get(scriptRef.getId());
```
Comment (pq) [inline src/io/flutter/coverage/ScriptManager.java]: Do we need to check for a null `scriptRef`?

File: src/io/flutter/coverage/ScriptManager.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        final int tokenPos = iterator.next();
+        final int column = Math.max(0, iterator.next() - 1);
+        result.put(tokenPos, Pair.create(line, column));
+      }
+    }
+
+    return result;
+  }
+
+  public Script getScriptFor(ScriptRef ref) {
```
Comment (pq) [inline src/io/flutter/coverage/ScriptManager.java]: Might `ref` be null?

File: src/io/flutter/coverage/ScriptManager.java [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+          }
+        }
+      }
+    }
+
+    return null;
+  }
+
+  @Nullable
+  private Library getLibrary(LibraryRef libraryRef) {
```
Comment (jacob314) [inline src/io/flutter/coverage/ScriptManager.java]: Consider using
 CompletableFuture<Library> getLibrary(LibraryRef instance)
from EvalOnDartLibrary also performs this task. Perhaps add a TODO to standardize to a single implementation going forward.

File: src/io/flutter/view/FlutterViewMessages.java [hunk 14]
```diff
@@ -39,9 +39,8 @@ public static void sendDebugActive(@NotNull Project project,
                                      @NotNull VmService vmService) {
     final MessageBus bus = project.getMessageBus();
     final FlutterDebugNotifier publisher = bus.syncPublisher(FLUTTER_DEBUG_TOPIC);
-    app.setVmService(vmService);
     assert(app.getFlutterDebugProcess() != null);
-    app.setPerfService(new PerfService(app.getFlutterDebugProcess(), vmService));
+    app.setVmServices(vmService, new PerfService(app.getFlutterDebugProcess(), vmService));
```
Comment (jacob314) [inline src/io/flutter/view/FlutterViewMessages.java]: Add a TODO to rename PerfService to VMService.

File: src/io/flutter/coverage/ScriptManager.java (hunk 12 above)
Comment (devoncarew) [reply src/io/flutter/coverage/ScriptManager.java]: I changed the param to be NotNull (and the return to be Nullable).

File: src/io/flutter/view/FlutterViewMessages.java (hunk 14 above)
Comment (devoncarew) [reply src/io/flutter/view/FlutterViewMessages.java]: done

File: src/io/flutter/coverage/FileCoverageInfo.java (hunk 5 above)
Comment (devoncarew) [reply src/io/flutter/coverage/FileCoverageInfo.java]: I'd like to keep this as is - I think this method (and the other `addUncovered()` method, and `addCovered`) can have knowledge of manipulating the `coveredLines` and `uncoveredLines` data structures.

**Pull request merged**